pkg/controller/daemon/update_test.go

```diff
@@ -117,6 +117,60 @@ func TestDaemonSetUpdatesPodsWithMaxSurge(t *testing.T) {
 	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
 }
 
+func TestDaemonSetUpdatesPodsNotMatchTainstWithMaxSurge(t *testing.T) {
+	_, ctx := ktesting.NewTestContext(t)
+
+	// Create DaemonSet with toleration
+	ds := newDaemonSetWithTolerations("foo")
+	manager, podControl, _, err := newTestController(ctx, ds)
+	if err != nil {
+		t.Fatalf("error creating DaemonSets controller: %v", err)
+	}
+
+	// add taint to one node
+	addNodesWithTaints(manager.nodeStore, 0, 5, 1, nil)
+	manager.dsStore.Add(ds)
+	expectSyncDaemonSets(t, manager, ds, podControl, 5, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	// surge is the controlling amount
+	maxSurge := 1
+	ds.Spec.Template.Spec.Containers[0].Image = "foo2/bar2"
+	ds.Spec.UpdateStrategy = newUpdateSurge(intstr.FromInt(maxSurge))
+	// remove toleration from DaemonSet for update
+	ds.Spec.Template.Spec.Tolerations = nil
+	manager.dsStore.Update(ds)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, maxSurge, 0, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, maxSurge, maxSurge, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, maxSurge, maxSurge, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, maxSurge, maxSurge, 0)
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, 0, 0)
+	markPodsReady(podControl.podStore)
+
+	clearExpectations(t, manager, ds, podControl)
+	expectSyncDaemonSets(t, manager, ds, podControl, 0, maxSurge, 0)
```

Comment:
so this would mean we got only 4 deletions, even though there should have been 5.

Can you please add tests that show a changing pods from a set of nodes Y to a set of nodes Z. Might be good to add some intersection between Y and Z for observing normal surge.

Eg you can test following transitions
- start DS on nodes Y
- move to nodes Z
- move to nodes Z+Y
- remove all tolerations = 0 pods running

I think this bug/feature fix requires an E2E test to show it is working properly and to show there are no extra pods hanging around after each of these transitions.

Reply:
I'll start working after the rollingUpdate fix is complete.

Reply:
@mochizuki875 I think it would be good to work on the test at the same time, as it can reveal bugs in the implementation

Reply:
I've fixed that test case and added new case which dose what you suggested.